### PR TITLE
Core julia support

### PR DESCRIPTION
**Pull request by jbendes**

### CODE REVIEW

**Comment by rdeits** [comment]
This is really exciting! I'd be happy to move from https://github.com/rdeits/LCMCore.jl to ZCM once Julia support comes in. 

I was wondering, though, why you chose to go make all of the Julia ZCM methods `::Function` fields inside the type? Doing so actually introduces a performance penalty (because `::Function` is an abstract type, see: https://docs.julialang.org/en/latest/manual/performance-tips/#Avoid-fields-with-abstract-type-1 ). For example:

```julia
julia> using BenchmarkTools

julia> mutable struct T1
         x::Int
         good::Function
         function T1(x)
           instance = new(x)
           instance.good = () -> instance.x > 5
           instance
         end
       end

julia> struct T2
         x::Int
       end

julia> good(t2::T2) = t2.x > 5
good (generic function with 1 method)

julia> t1 = T1(6)
T1(6, #1)

julia> t2 = T2(6)
T2(6)

julia> @btime ($t1).good()
  12.691 ns (0 allocations: 0 bytes)
true

julia> @btime good($t2)
  1.604 ns (0 allocations: 0 bytes)
true
```

so the `::Function` field design will introduce a performance penalty (and a type-instability) for every ZCM method. That's why that style is pretty uncommon in Julia libraries. 

Anyway, I don't mean to be critical--this is really exciting, and I'm really grateful that you've put this together. I'm happy to talk more about Julia code style if you're interested.

**Comment by jbendes** [comment]
No I am very appreciative of the feedback!! I'm actually not a huge Julia user myself so I was unaware of this. How would you have set up those functions? Any other feedback here? The pullreq is still pending review before merge so if you were interested in giving it an examination and providing feedback I'd be very grateful :)

**Comment by jbendes** [comment]
Hey @rdeits, working on switching over to the method based function dispatch as you suggested. Have it working with zcm core, but there is an issue when attempting to resolve the `encode` and `decode` functions associated with a type on publish and subscribe. Could the issue be because the `encode` and `decode` functions aren't inside `zcm.jl`? Don't have enough Julia knowledge to know how to properly do it. They are inside the type's file but seemingly aren't exposed inside `zcm.jl`. If you compile the repo and the examples, you'll see the two files in question in `zcm/julia/zcm.jl` and `examples/build/types/example_t.jl`.

**Comment by rdeits** [comment]
Sure, I'll take a look.

**Comment by rdeits** [comment]
One thing I've found as I'm trying to build this is that you're including `<julia/uv.h>` which assumes that Julia is installed somewhere standard. But the [Julia instructions](https://julialang.org/downloads/platform.html#generic-linux-binaries) don't recommend installing Julia systemwide nor do they recommend the version from `apt-get` (which is not maintained by the Julia developers and is generally very out-of-date). Is there some way for me to tell `waf` to include my Julia header directory? I tried `-I/home/rdeits/apps/julia-0.6.0/include` but it looks like `waf` doesn't use `-I`

**Comment by jbendes** [comment]
@rdeits Can you tell we're not julia masters? :)
For now, a hot fix to get you compiling would be to modify the wscript in `zcm/julia`. Simply add `'includes' =  '/path/to/dir'` to the `shlib` directive. How would you recommend enabling julia build without knowing where julia is installed? We could change the --use-julia to require a path but i'm not fond of that solution

**Comment by rdeits** [comment]
If the Julia executable is available on the user's path, then the correct include directory can be found with:

```bash
$ julia -E "abspath(JULIA_HOME, Base.INCLUDEDIR)"
```

Adding that folder to the includes should work for most users. 

Alternatively, some users may have the JULIA_HOME environment variable set. In that case, the correct path would be `$JULIA_HOME/../include`. 

I would suggest modifying the wscript to first look for the `$JULIA_HOME` environment variable, and then, if it's not found, try running `julia  -E "abspath(JULIA_HOME, Base.INCLUDEDIR)`.

**Comment by rdeits** [comment]
Also, I'm happy to help make these changes, but it'll be a bit difficult since my laptop is OSX-only.

**Comment by rdeits** [comment]
Ok, I see the issue with encode/decode. You're correct that it's because the functions aren't defined inside ZCM.jl. In essence, what you're doing now is creating an entirely new function called `encode()` for each type, whereas what you want is to create a new *method* for the single ZCM.encode() function for each type. 

My suggestion would be:

* In ZCM.jl, you can create function stubs like this:

```julia
function encode end
function decode end
```

and so on. This will create a function with no methods (so attempting to call it will give an appropriate error). 

Then, in each zcmtype-generated file, you want to do:

```julia
import ZCM
function ZCM.encode(msg::example_t)
   <function body here>
end
```

this will ensure that all of your `encode` methods belong to the central ZCM.encode function. 

I also noticed that your `decode()` function takes just a UInt8 buffer as its only argument. That means that the function signature itself doesn't distinguish what type you're trying to decode into. I would suggest changing the signature to:

```
function decode(::Type{example_t}, data::Vector{UInt8})
  <function body>
end
```

And users would call it with `decode(example_t, data)`. The advantage is that this version allows all of the `decode()` methods to likewise belong to the same single `decode()` function, but their argument types specify how the data is supposed to be decoded. This approach (with a type as the first argument) is very common in Julia, as in `zeros(Int, 5)`.

**Comment by rdeits** [comment]
By the way, my colleague @tkoolen already did a lot of work on making a nice lcmtype interface in Julia: https://github.com/rdeits/LCMCore.jl/blob/master/src/lcmtype.jl The only piece that's missing from his implementation was parsing the `.lcm` file format in order to automatically generate the Julia data structures. But perhaps we could combine your parsing with his design for the Julia message types, since he's already done a lot of work on making them idiomatic and performant.

**Comment by olsoni** [comment]
@rdeits I just pushed up with the modification to our wscript to read in the julia include path by using the `julia -E "abspath(JULIA_HOME, Base.INCLUDEDIR)"`. I didn't add the use of the `JULIA_HOME` environment variable yet because I was curious about how that would ever be different (and correct to use). It's important that we use to the same libuv headers that julia itself uses when running. Is there ever a case where the internal `JULIA_HOME` directory in the julia repl is *not* the location of the libuv that is used by julia?

**Comment by jbendes** [comment]
Hey @rdeits, take another look if you have a second. Just implemented what you suggested and it seems to be working again!

**Comment by rdeits** [comment]
I think it's totally reasonable to expect `julia` to be on the path. That said, I believe it is always safe to trust the `JULIA_HOME` environment variable if it's set. The reason I think that is because if you set `JULIA_HOME` to a directory which is *not* where `julia` is installed, then Julia itself will fail:

```bash
# With no JULIA_HOME, everything is fine:
rdeits@hostname-21: ~/Projects/zcm
$ julia -e ""

# if JULIA_HOME is set, it should match the `bin` directory of Julia:
rdeits@hostname-21: ~/Projects/zcm
$ JULIA_HOME=/home/rdeits/apps/julia-0.6.0/bin julia -e ""

# Setting JULIA_HOME to a nonexistent directory will fail:
rdeits@hostname-21: ~/Projects/zcm
$ JULIA_HOME=/nonexistent/path julia -e ""
ERROR: could not load library "/nonexistent/path/../lib/julia/sys.so"
/nonexistent/path/../lib/julia/sys.so: cannot open shared object file: No such file or directory

# Even setting it to the wrong Julia version will fail:
rdeits@hostname-21: ~/Projects/zcm
$ JULIA_HOME=/home/rdeits/apps/julia-0.5.0/bin julia -e ""
ERROR: could not load library "/home/rdeits/apps/julia-0.5.0/bin/../lib/julia/sys.so"
/home/rdeits/apps/julia-0.5.0/bin/../lib/julia/sys.so: cannot open shared object file: No such file or directory
```

So Julia is already enforcing that JULIA_HOME is sane.

**Comment by rdeits** [comment]
Ok, excellent! I'll take a look at it in more detail as soon as I can (it may not be until tomorrow, though). By the way, I made a minor change to improve loading ZCM.jl in https://github.com/ZeroCM/zcm/pull/164 (it's a PR against this branch, not master).

**Comment by olsoni** [comment]
just merged #164 (though I rebased it first so I might have bungled the github pullreq to say closed instead of merged)

File: zcm/julia/ZCM.jl [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+end
+
+type Sub
+end
+
+type UvSub
+end
+
+type RecvBuf
+    recv_utime ::Int64;
```
Comment (tkoolen) [inline zcm/julia/ZCM.jl]: Semicolons aren't necessary (anywhere) and are kind of nonstandard.

**Comment by rdeits** [comment]
Ok, this is definitely getting better :). A few more suggestions:

* It's typical to put all of the import and export statements at the top of the file, so they're easy to see at a glance. You can combine multiple imports and exports into a single statement:

```
import ZCM: encode, decode
export foo, bar, baz
```

* The way you've defined `ntoh(arr::Array{T})` is not recommended. It's technically "type piracy" because you're changing the behavior of someone else's function (`ntoh`) on someone else's datatype (`Array`). Fortunately, it's also totally unnecessary. You can just call `ntoh.(arr)` to apply `ntoh` to every element of `arr` (creating a new array for the result). If you want to do the conversion inplace, you can do `arr .= ntoh.(arr)`. 

* The `decode_one`, `_get_hash_recursive`, and `getHash()` functions should also take a `::Type{example_t}` argument. Currently, trying to `include()` two different generated zcmtypes will cause the second one's `getHash()` to clobber the first one's. This is basically the same problem that `encode` and `decode` had before, and the solution is the same: stub them in ZCM.jl and extend them with `function ZCM.getHash(::Type{example_t}) ... end` in the generated code. 

* The generated zcmtypes use `takebuf_array`, which is deprecated in Julia v0.6 and will be gone in v0.7 later this year. The updated syntax is `take!(buf)`. If you want to maintain compatibility with both versions, you can do:

```julia
@static if VERSION < v"0.6.0-"
  _takebuf_array(buf) = takebuf_array(buf)
else
  _takebuf_array(buf) = take!(buf)
end
```

and then call the new `_takebuf_array()` function.

File: zcm/julia/ZCM.jl [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+end
+
+# RRR: make variable naming consistent
+# RRR: improve orgainzation and make header comments more visible
+# Exported Objects and Methods
+type RecvBuf
+    rbuf  ::Ptr{Native.RecvBuf};
+    utime ::Int64;
+    data  ::Array{UInt8};
```
Comment (rdeits) [inline zcm/julia/ZCM.jl]: This should be `Array{UInt8, 1}` or (equivalently, because it's just an alias), `Vector{UInt8}`. Currently, this field can hold an array with any number of dimensions, but I'm pretty sure you just have 1-D vectors. Changing the field will make its type concrete and improve performance.

File: zcm/julia/ZCM.jl [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        # destruction of the underlying Native.RecvBuf pointer is handled by the C lib internally
+
+        return instance;
+    end
+end
+export RecvBuf;
+
+type __zcm_handler
+    handler::Any;
+    usr::Any;
```
Comment (tkoolen) [inline zcm/julia/ZCM.jl]: Instead of having `handler` and `usr` be of type `Any`, perhaps it makes more sense to parameterize `__zcm_handler` on the types of `handler` and `usr`, i.e.:
```julia
type __zcm_handler{H, U}
    handler::H
    usr::U
end
```
`Sub` would then also be parameterized on the concrete `__zcm_handler` subtype.

File: zcm/julia/ZCM.jl [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+    # handler must have the function or functor prototype :
+    #   (rbuf::RecvBuf, channel::String, msg::MsgType, usr)
+    __zcm_handler(handler, usr) = new(handler, usr);
```
Comment (tkoolen) [inline zcm/julia/ZCM.jl]: This constructor is already generated by default.

File: zcm/julia/ZCM.jl [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    end
+
+    function Zcm() return Zcm(""); end
+end
+export Zcm;
+
+function encode end
+function decode end
+
+function good(zcm::Zcm)
```
Comment (tkoolen) [inline zcm/julia/ZCM.jl]: `isgood` would be more standard Julia naming, but I understand if you want to make things as close as possible to the other languages.

File: zcm/julia/ZCM.jl (hunk 4 above)
Comment (rdeits) [inline zcm/julia/ZCM.jl]: This constructor is unnecessary. In addition, I'm not sure that the subscription interface is being handled in the best way. Particularly, passing the `usr` field around as a pointer and then doing `unsafe_pointer_to_objref` is scary and unnecessary. Instead, I would strongly suggest just asking users to pass a closure over whatever user variables they're interested in. That's what I do in https://github.com/rdeits/LCMCore.jl 

So, rather than a user doing (pseudocode):

```julia
subscribe(zcm, "FOO", foo_t, callback, usr)
```

they would just construct a callback that closes over their user-data and subscribe that:

```
callback = (buf, channel) -> myfunction(buf, channel, usr)
subscribe(zcm, "FOO", foo_t, callback)
```

That way you're letting Julia do the work of managing the user data and ensuring it doesn't get garbage collected. 

In fact, I've been pretty happy with the subscription system in LCMCore.jl, and I think it could probably be ported entirely into ZCM. Want me to give it a try?

File: zcm/julia/ZCM.jl [hunk 6]
```diff
@@ -0,0 +1,334 @@
+module ZCM
+
+# RRR / Note: Julia requires that the memory layout of the C structs is consistent
+#             between their definitions in zcm headers and this file
+
+# C ptr types for types that we don't need the internals of
+module Native
+
+type Zcm
```
Comment (tkoolen) [inline zcm/julia/ZCM.jl]: I think a lot of these `type`s could/should be `immutable`s (though I'm not 100% sure about what's required by the `ccall` interface; @rdeits has more experience with that).

Also, in Julia 0.6 (and future 1.0), `type` will be `mutable struct` and `immutable` will be `struct`.

File: zcm/julia/ZCM.jl [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+export publish;
+
+# TODO: force msg to be derived from our zcm msg basetype
+function publish(zcm::Zcm, channel::AbstractString, msg)
+    data = encode(msg)
+    return publish(zcm, channel, data, UInt32(length(data)));
+end
+export publish;
+
+function flush(zcm::Zcm)
```
Comment (tkoolen) [inline zcm/julia/ZCM.jl]: There is a `Base.flush` that you probably should be extending here (i.e. replace `flush` with `Base.flush`).

File: zcm/julia/ZCM.jl (hunk 4 above)
Comment (rdeits) [reply zcm/julia/ZCM.jl]: Also, I think that the Zcm object should actually contain the list of subscriptions. The reason is that if the `Sub` returned by `subscribe()` goes out of scope, the ZCM subscription still exists. That can result in the handler being called after the Julia `Sub` has been freed, which will crash. That's also something I can port over from LCMCore.jl

File: zcm/julia/ZCM.jl [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+    end
+end
+export LogFile;
+
+function good(lf::LogFile)
+    return lf.eventLog != C_NULL;
+end
+export good;
+
+function readNextEvent(lf::LogFile)
```
Comment (tkoolen) [inline zcm/julia/ZCM.jl]: Lower case and underscored is standard in Julia, but again I understand if you want to make things have the same name between languages.

File: zcm/julia/ZCM.jl [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+    utime   ::Int64;
+    channel ::String;
+    data    ::Array{UInt8};
+
+    function LogEvent(event::Ptr{Native.EventLogEvent})
+        instance  = new();
+        instance.event = event;
+
+        if (event != C_NULL)
+            loadedEvent = unsafe_load(event);
```
Comment (tkoolen) [inline zcm/julia/ZCM.jl]: Lower case and underscored is standard in Julia.

**Comment by tkoolen** [comment]
Is there any particular reason for writing and this against Julia 0.5 as opposed to 0.6 BTW?

File: zcm/julia/ZCM.jl (hunk 6 above)
Comment (rdeits) [reply zcm/julia/ZCM.jl]: Anything that has a finalizer has to be immutable, otherwise yes I agree.

Comment (rdeits) [reply zcm/julia/ZCM.jl]: Errr, nope, exactly wrong. Anything with a finalizer has to be *mutable*

File: zcm/julia/ZCM.jl (hunk 3 above)
Comment (rdeits) [reply zcm/julia/ZCM.jl]: Agreed, but this type may go away if I port in the code from LCMCore

**Comment by jbendes** [comment]
Very excited to get to this! Isaac and I will hopefully get back to this tonight/tomorrow. The short answer at to why 0.5 vs 0.6 is because that's what apt-get gives you :)

**Comment by tkoolen** [comment]
I see. Yeah, as Robin mentioned [earlier](https://github.com/ZeroCM/zcm/pull/163#issuecomment-342556085), `apt` is not the recommended way to get Julia. I would recommend targeting 0.6.1 (i.e., the latest release) and additionally using `struct` instead of `immutable` and `mutable struct` instead of `type` throughout your code in anticipation of Julia 1.0.

**Comment by jbendes** [comment]
Definitely makes sense! Just finished reading over your guy's comments. Seems like your pullreq #165 addresses a bunch of the suggestions from above so I'll just go review that now.

**Comment by tkoolen** [comment]
I'm having trouble building this locally.

Build output:
```
twan@tk-loco:~/code/zcm$ ./waf configure --use-julia
Setting top to                           : /home/twan/code/zcm 
Setting out to                           : /home/twan/code/zcm/build 
Checking for 'gcc' (C compiler)          : /usr/lib/ccache/gcc 
Checking for 'g++' (C++ compiler)        : /usr/lib/ccache/g++ 
Checking for program 'julia'             : /opt/julia-0.6/bin/julia 
ZCM Version: 1.0.0 

Dependency Configuration: 
    C/C++                Enabled 
    Java                 Disabled 
    NodeJs               Disabled 
    Python               Disabled 
    Julia                Enabled 
    ZeroMQ               Disabled 
    Elf                  Disabled 
    Third Party          Disabled 

Transport Configuration: 
    ipc                  Disabled 
    inproc               Disabled 
    udpm                 Disabled 
    serial               Disabled 

Type Configuration: 
    hash-typename        Enabled 
    hash-member-names    Disabled 

Dev Configuration: 
    Clang                Disabled 
    CxxTest              Disabled 
 
'configure' finished successfully (0.787s)
twan@tk-loco:~/code/zcm$ ./waf build
Waf: Entering directory `/home/twan/code/zcm/build'
[21/45] Compiling zcm/julia/uv_zcm_msg_handler.cpp
[22/45] Compiling config/zcm.pc.in
[41/45] Creating build/zcm.gitid
../zcm/julia/uv_zcm_msg_handler.cpp:6:22: fatal error: julia/uv.h: No such file or directory
compilation terminated.

Waf: Leaving directory `/home/twan/code/zcm/build'
Build failed
 -> task in 'zcmjulia' failed (exit status 1): 
        {task 140223478858128: cxx uv_zcm_msg_handler.cpp -> uv_zcm_msg_handler.cpp.1.o}
['/usr/lib/ccache/g++', '-fPIC', '-std=c++11', '-fPIC', '-pthread', '-Wall', '-Werror', '-Wno-unused-function', '-O3', '-Wno-unused-local-typedefs', '-I.', '-I..', '-I..', '-Izcm/julia/"/opt/julia-0.6/include"', '-I../zcm/julia/"/opt/julia-0.6/include"', '-D_LARGEFILE_SOURCE', '-D_FILE_OFFSET_BITS=64', '-DUSING_CPP', '-DUSING_JULIA', '-DUSING_OPT', '-DENABLE_TYPENAME_HASHING', '../zcm/julia/uv_zcm_msg_handler.cpp', '-c', '-o/home/twan/code/zcm/build/zcm/julia/uv_zcm_msg_handler.cpp.1.o']
```

Any ideas?

**Comment by jbendes** [comment]
Can you run this and paste the output?
`/opt/julia-0.6/bin/julia -E "abspath(JULIA_HOME, Base.INCLUDEDIR)"`

**Comment by tkoolen** [comment]
Most recent push fixed it. Thanks!

**Comment by rdeits** [comment]
There's still at least one issue that I know of here, which is that including two different message types will break things due to each message type defining a separate `getHash()` function.

**Comment by rdeits** [comment]
Another thing from discussion with @tkoolen is that it would be nice to have ZCM type packages correspond to Julia modules. That's analogous to how:

```c
package foo;

struct bar_t {
}
```

creates a Python package called `foo` containing the `bar_t` class. 

In Julia, we can do this by having each type generate a `.jl` file for the type, and the create a module that `include()`s each type file. 

To be concrete, I'm proposing that if we have the `foo` package as above, we would generate this folder structure:

```
build/
	types/
		foo.jl
		foo/
			bar_t.jl
```

where `foo.jl` would look like the following:

```julia
module foo

include("foo/bar_t.jl")

end
```

and `bar_t.jl` would look like:

```julia
type bar_t # not foo_bar_t
  ...
end
```

With that change, users could add the `build/types` folder to `JULIA_LOAD_PATH` and then do:

```
julia> import foo
julia> foo.bar_t()
```

**Comment by rdeits** [comment]
rdeits/fix-multi-types adds a (broken) multi_type_test.jl to demonstrate the issue with multiple definitions of getHash() and co. I'll try to fix it now.

**Comment by jbendes** [comment]
Will definitely take a look at that test. As for the package, we decided that support for packages was limited by julia's inability to split modules over multiple files. You had to have a single entry point. So we opted for the c-style underscored package naming. Did you look at how packages function currently?

**Comment by rdeits** [comment]
Actually, it's quite common for julia modules to be split over multiple files (that's the primary thing that `include()` is used for). You can just do:

```julia
module Foo
include("bar.jl")
include("baz.jl")
end
```

and the contents of `bar.jl` and `baz.jl` will be included at the top level in module Foo.

**Comment by rdeits** [comment]
You do need a single file that actually defines the module, but it's exactly analogous to Python's `__init__.py`, which zcm-gen already knows how to construct.

**Comment by tkoolen** [comment]
Actually, you can also add stuff to an existing module. Say you have a file `bar.jl` with contents
```julia
struct Bar
    a::Int
end
```
Then you can do
```julia
julia> module M end
M

julia> eval(M, :(include("bar.jl")))

julia> M.Bar(1)
M.Bar(1)
```

**Comment by rdeits** [comment]
Oh yeah, that's even better :)

**Comment by jbendes** [comment]
Yea i like that a lot. We didn't think that was possible. Though.......that syntax though........haha

**Comment by rdeits** [comment]
Actually, the `eval()` trick may not be super helpful, because you still need some single file that tells Julia to do the appropriate `eval()`s. But at that point, the single file might as well just be `foo.jl` and might as well just directly `include("foo/bar.jl")` (as in my example above).

**Comment by jbendes** [comment]
Is there a way to know if a module exists? I'm going to put this at the top of each zcmtype
```
eval(<thistype_package>, :(include("<thistype.jl>")))
```

**Comment by rdeits** [comment]
You can check it with `isdefined(Main, :thistype_package)`, but I'm not convinced this is the right thing to do. 

More generally, in Python I can currently do:

```python
import thistype_package
thistype_package.thistype()
```

or 

```python
from thistype_package import thistype
thistype()
```

Likewise, I suggest having zcm-gen create a single Module that `include()`s each type's definition in the same way that `__init__.py` works. Then Julia users can do:

```julia
import thistype_package
thistype_package.thistype()
```

or 

```julia
import thistype_package: thistype
thistype()
```

Rather than directly `include()`-ing files. The problem with what you're suggesting is that once we start conditionally putting types into packages and asking users to `include()` files directly, it's very easy for them to end up with extremely weird situations where two types might belong to the same package, but only one actually lives in the Julia module because of the way or order in which it was imported.

**Comment by jbendes** [comment]
Is that `eval` syntax 0.5? I am getting this error
```
ERROR: LoadError: LoadError: syntax: invalid character literal
```
the code is
```
module test_package end
eval(test_package, :(include('packaged_t.jl')))
```

**Comment by rdeits** [comment]
It might not be v0.5 compatible. I think I'm pretty close to getting the python-style version working, which won't require eval at all.

**Comment by jbendes** [comment]
Also regarding the organization of packages, I don't mind additionally having a global package file but I would rather not need to import an entire package. I think we should maintain ability to import individual parts of packages. Take a look at the commit I just pushed

**Comment by tkoolen** [comment]
> `'packaged_t.jl'`

That should be `"packaged_t.jl"`. Unlike in Python, you can only use double quotes for strings. Single quotes are for characters.

**Comment by rdeits** [comment]
I agree that this works, but I don't think it's a good idea. You're modifying the behavior of the program based on some transient global state (whether a module is defined at the particular moment that the file was included). This means that, for example, re-arranging the order of imports/includes will result in fundamentally different program behavior. 

I think I can solve this and give both of us what we want. If we use the folder structure I outlined above:

```
build/
  types/
    foo.jl
    foo/
      bar.jl
```

where `foo.jl` contains:

```julia
module foo 
include("foo/bar.jl")
end
```

and `bar.jl` does *not* have any of the `isdefined()` or `eval()` code, then:

* A user who wants to import the package can do `import foo`
* A user who wants just a single type without the package can do `include("build/types/foo/bar.jl")`

With no possibility of program behavior depending on the order of imports.

**Comment by jbendes** [comment]
I don't quite see the issue you're describing. Zcm types have recursive types in them. So let's say I have type `recursive_t` which contains an `example_t` inside of it. At the top of `recursive_t.jl` there would be a line `include("example_t.jl")`. Now if we introduce packages into this. Both types are in package `test`. That means that if the files that the types are in dont contain their package information, `include("recursive_t.jl")` would include `recursive_t` and `example_t` as types not in packages.

**Comment by jbendes** [comment]
Also, I fixed the single quotes thing and now it's saying this:
```
ERROR: LoadError: LoadError: syntax: module expression not at top level
```
for this code
```
if !isdefined(Main, :test_package)                                                                   
    module test_package end                                                                          
end                                                                                                  
eval(test_package, :(include("packaged_t.jl"))) 
```

**Comment by rdeits** [comment]
Ah, yes, that's true. Module declarations must be at the top level and thus can't be conditional (modules are fundamental to the way Julia types are organized, so they can't be defined at run-time except through metaprogramming magic). This is Julia trying to tell you that conditionally defining a module is a bad idea. 

As for the situation you're describing, I agree that it gets messy, but I think this is more of a reason to *always* use modules and *always* use the import machinery built into Julia. Working around it with conditional modules and `eval()` still seems like a bad idea to me. 

But I think I don't actually understand your use-case entirely. You've said that you don't want to always import an entire package, but can you tell me why? That is, what's wrong with `import packagename: typename`?

**Comment by jbendes** [comment]
Yea that does make sense. It's a bit weird that you'd still be able to include individual types but not get them in their package. Also are we introducing a namespace collision by naming the package `test.jl` at the same level as `test/`? What if i also had a type named `test`?

**Comment by rdeits** [comment]
Totally agreed :). 

Good point. It's generally not a problem, but I can imagine a corner case. In particular, when you do `import foo`, Julia looks on your LOAD_PATH for `foo.jl` or `foo/src/foo.jl`. The only way we'd see a collision is if you either have a ZCM type with the same name as it's package (having a module contain a type with the same name as its parent module is not allowed) or if some trickster decided to make a package called `src` and a type called `foo` in addition to a package called `foo`. 

I think this is pretty straightforward to work around by just modifying the package structure a bit. 

By the way, I think I'm close to a working prototype of EmitJulia using modules. It already works for the existing tests, and I'm trying to bring in recursive types now.

**Comment by jbendes** [comment]
Fantastic can't wait to see it

**Comment by jbendes** [comment]
Are you making it so it only works if you specify all the types in one instance of `zcm-gen`? Ie what if i call `zcm-gen` on two different types both belonging to the same package?

**Comment by rdeits** [comment]
Uh, great question. I think I'm doing it correctly (like in python, I'm reading the existing package file (`__init__.py` or `modulename.jl`) and adding to, rather than replacing, its contents. But I've probably missed some edge cases...

**Comment by rdeits** [comment]
Okay, prototype implementation is available in https://github.com/rdeits/zcm/tree/package-modules 

What works:

* packaged types get put into a module with the package name
* `import packagename` and `import packagename: typename` work
* recursive types and types which depend on types from other packages work

What doesn't work:
* nested packages don't work at all yet
* type definition files can get included multiple times, which results in warnings

That last one is tricky. The problem is that Julia requires that a type be defined before it's used as a field of a struct, so 

```
type A
  x::B
end

type B
  y::Int
end
``` 

is not OK. C++ gets around this with `#ifndef` guards, but I'd rather not go that route for Julia. I'll have to think about whether there's a better solution. 

Meanwhile, if you take a look at #166 it has some minor but important fixes that should be applied regardless of how we handle packaging.

**Comment by jbendes** [comment]
Hey @rdeits take a look at a16157af847be83a205e02cf33350403e63d9590 to see if you agree with my edits

**Comment by rdeits** [comment]
Looks good to me

**Comment by rdeits** [comment]
@jbendes Ok, I finally feel like I've got my head wrapped around creating Julia modules for the ZCM types. I still think it's the right thing to do, but actually implementing it might take a day or two. So I'm going to open a new issue to create a place to discuss it without filling up this github thread. 

If you want to go ahead and merge this into master without ZCM packages being Julia modules, I think that's fine. The switch from underscores to proper modules will be a breaking change, but one that I hope can happen in just a day or two, so I hope that's alright ;-)

**Comment by jbendes** [comment]
Okay little endian support now available

**Comment by olsoni** [comment]
Admittedly after *much* longer than I had anticipated, this branch is finally ready for some more eyes. 

Summary of features that are now working:
- Type generation using the zcmtype packages to organize the module structure. 
- publish, subscribe, handlers, typed handlers, handlers with extra arguments (closure-esk)
- support for pause, flush, resume
- reading and writing logfiles

Things that could use an extra look or a little polish
- Interface with the c library uv functions isn't as hardened as I'd like it. If you have an assertion or other error during a handler, you often can freeze all of julia and need to kill -9 it from another terminal
- Some parts might need to be updated to Julia 0.6 syntax
- There are a few places in calling the C interface where I'm not 100% sure I'm treating pointers in a way that will work properly with the Julia garbage collector. I haven't run into any problems in my tests, but I haven't run any tests at large scale, so I'm not confident that it's 100% right.

**Comment by olsoni** [comment]
Would you like to somehow leave a trimmed down version of that example in the zcmgen comment? Presumably you don't mean "good to merge yet" as you still have those too files left, you just mean that section is fine for now?

**Comment by jbendes** [comment]
Correct. The stuff I looked over is fine. And yea that example in its entirety I think is fine to leave there haha
